### PR TITLE
refactor: flatten MariadbGTIDSet to map[uint32]*MariadbGTID (one entry per domain)

### LINKED ISSUE #1123: MariadbGTIDSet: data model stores multiple server IDs per domain; Contain() comparison is incorrect across server IDs
## Background

Identified during the fix in #1122: `MariadbGTIDSet` uses `map[uint32]map[uint32]*MariadbGTID` (domain_id → server_id → GTID), which stores multiple server entries per domain. However, MariaDB only tracks **one position per domain_id** in `gtid_slave_pos` and `MASTER_GTID_POS`.

### 1. Data model allows multiple server IDs per domain

The nested `map[serverID]*MariadbGTID` per domain can accumulate entries from different servers. While #1122 fixes `AddSet()` to clear old entries on server change, the underlying data model still permits it. Ideally, the structure should enforce one entry per domain.

**Introduced in:** PR #852 (`034c7ab` — "separate serverID of Mariadb GTID set", 2024-03-14)
**Previous behavior:** PR #272 (`cb167c0`) used a flat `map[uint32]*MariadbGTID` with one entry per domain.

### 2. `Contain()` is incorrect across server IDs

`MariadbGTID.Contain()` checks if `DomainID` matches and `SequenceNumber >= other.SequenceNumber`. However, if the two GTIDs have different `ServerID` values (e.g., after a primary failover), comparing sequence numbers is not meaningful — sequences are per-server, not globally ordered across servers within a domain.

As @dveeden noted: *"For MariaDB you can't [tell if a specific GTID was executed] as the current GTID position may have a different serverId than the GTID you're comparing against."*

### 3. Naming: `MariadbGTIDSet` is a position, not a set

Unlike MySQL's `GTIDSet` which tracks ranges of executed transactions, MariaDB's GTID represents a **position** in each replication stream. The current naming can be misleading.

## Suggested improvements

1. Consider flattening the map back to `map[uint32]*MariadbGTID` (one entry per domain)
2. Update `Contain()` to account for server ID differences
3. Consider renaming or adding documentation to clarify the position-vs-set distinction

cc @dveeden @lance6716

### PR DESCRIPTION
**Pull request by triplefighter7619**
Fixes #1123

## Problem

`MariadbGTIDSet` used `map[uint32]map[uint32]*MariadbGTID` (domain_id → server_id → GTID),
which allowed multiple server entries per domain. This is incorrect for MariaDB, which tracks
exactly **one GTID position per domain_id** in `gtid_slave_pos` and `MASTER_GTID_POS`.

This structure was introduced in PR #852 (`034c7ab`). Before that, PR #272 (`cb167c0`) used a
flat `map[uint32]*MariadbGTID` which correctly enforced one entry per domain.

### `mariadb_gtid.go`

- **`MariadbGTIDSet.Sets`**: flattened from `map[uint32]map[uint32]*MariadbGTID` to `map[uint32]*MariadbGTID`
- **`AddSet()`**: simplified — if domain exists, call `forward()` to update server ID + sequence; otherwise add new entry. The `forward()` method already handles server ID changes (as happens during primary failovers)
- **`String()`**, **`Encode()`**, **`Clone()`**, **`Equal()`**, **`Contain()`**: all simplified from nested iteration to single-level map iteration
- Removed unused `"bytes"` import

### `mariadb_gtid_test.go`

- Updated all test expectations from `map[uint32]map[uint32]string` to `map[uint32]string`
- Added `TestMariaDBGTIDSetSameDomainDifferentServer` with comprehensive coverage:
  - Basic server replacement (primary failover scenario)
  - Multi-domain isolation (only affected domain changes)
  - Direct `AddSet()` call (as `binlogsyncer` does)
  - Equality after forward
  - Clone independence
- Removed unused `"slices"` import

### No changes to `binlogsyncer.go`

`AddSet(&event.GTID)` passes `*MariadbGTID` — same signature, works with the flat map.

## Context

As discussed in #1122 and noted by @dveeden and @lance6716, MariaDB's GTID is a **position**
(one per domain), not a **set** (like MySQL's executed GTID set). The flat map structure
enforces this at the type level, making it impossible to accidentally accumulate multiple
server entries under the same domain.

### CODE REVIEW

File: mysql/mariadb_gtid.go [hunk 1]
```diff
@@ -121,23 +122,20 @@ func ParseMariadbGTIDSet(str string) (GTIDSet, error) {
...
-		serverSets[gtid.ServerID] = gtid
-	} else {
-		err := o.forward(gtid)
-		if err != nil {
+	if existing, ok := s.Sets[gtid.DomainID]; ok {
+		if err := existing.forward(gtid); err != nil {
 			return errors.Trace(err)
 		}
+	} else {
+		s.Sets[gtid.DomainID] = gtid
```
[the remark this answers is not in the document]
Comment (triplefighter7619) [reply mysql/mariadb_gtid.go]: Implemented this

File: mysql/mariadb_gtid.go [hunk 2]
```diff
@@ -174,29 +170,24 @@ func (s *MariadbGTIDSet) String() string {
...
-			sep = ","
-		}
+	for _, gtid := range s.Sets {
+		buf.WriteString(sep)
+		buf.WriteString(gtid.String())
+		sep = ","
 	}
 
-	return buf.Bytes()
+	return []byte(buf.String())
```
[the remark this answers is not in the document]
Comment (triplefighter7619) [reply mysql/mariadb_gtid.go]: Implemented this

**Review by dveeden — APPROVED** [review]

**Comment by triplefighter7619** [comment]
@lance6716 can you also take a look at this

**Comment by lance6716** [comment]
Sorry I'm busy these days. Maybe I'll have more spare time ~5 days later. @dveeden free to merge for now or request some AI review

**Comment by triplefighter7619** [comment]
@dveeden can you please merge this PR?

File: mysql/mariadb_gtid_test.go [hunk 3]
```diff
@@ -92,14 +91,16 @@ func TestMariaDBForward(t *testing.T) {
...
-		{"0-1-1", map[uint32]map[uint32]string{0: {1: "0-1-1"}}, []string{"0-1-1"}, false},
-		{"", nil, []string{""}, false},
-		{"0-1-1,1-2-3", map[uint32]map[uint32]string{0: {1: "0-1-1"}, 1: {2: "1-2-3"}}, []string{"0-1-1,1-2-3", "1-2-3,0-1-1"}, false},
-		{"0-1--1", nil, nil, true},
+		{"0-1-1", map[uint32]string{0: "0-1-1"}, "0-1-1", false},
+		{"", nil, "", false},
+		{"0-1-1,1-2-3", map[uint32]string{0: "0-1-1", 1: "1-2-3"}, "0-1-1,1-2-3", false},
+		{"0-1--1", nil, "", true},
+		// Same domain, different server — last one wins (forward replaces)
+		{"0-1-1,0-2-2", map[uint32]string{0: "0-2-2"}, "0-2-2", false},
```
Comment (dveeden) [inline mysql/mariadb_gtid_test.go:111]: Might be good to test for a non-numeric domain-id (`x-1-1`) which should result in an error.

File: mysql/mariadb_gtid.go [hunk 4]
```diff
@@ -125,25 +123,18 @@ func ParseMariadbGTIDSet(str string) (GTIDSet, error) {
...
-		}
-	} else if o, ok := serverSets[gtid.ServerID]; !ok {
-		slog.Warn("replacing server entries for domain", slog.Uint64("domainID", uint64(gtid.DomainID)), slog.Any("new", gtid))
-		clear(serverSets)
-		serverSets[gtid.ServerID] = gtid
-	} else {
-		err := o.forward(gtid)
-		if err != nil {
+	if existing, ok := s.Sets[gtid.DomainID]; ok {
+		if err := existing.forward(gtid); err != nil {
```
Comment (dveeden) [inline mysql/mariadb_gtid.go:133]: Looks like we don't have testing for a GTID that causes `forward()` to return an error. This can happen if the domain-id is different.

I think this can also happen when `Update()` calls `AddSet()`

File: mysql/mariadb_gtid.go
Comment (dveeden) [inline mysql/mariadb_gtid.go:191]: I think we are not testing this.

File: mysql/mariadb_gtid.go [hunk 5]
```diff
@@ -219,21 +195,14 @@ func (s *MariadbGTIDSet) Equal(o GTIDSet) bool {
 		return false
 	}
 
-	for domainID, set := range other.Sets {
-		serverSet, ok := s.Sets[domainID]
+	for domainID, gtid := range other.Sets {
+		ourGTID, ok := s.Sets[domainID]
 		if !ok {
 			return false
```
Comment (dveeden) [inline mysql/mariadb_gtid.go:201]: I think we are not testing this

File: mysql/mariadb_gtid.go [hunk 6]
```diff
@@ -246,17 +215,13 @@ func (s *MariadbGTIDSet) Contain(o GTIDSet) bool {
 		return false
```
Comment (dveeden) [inline mysql/mariadb_gtid.go:215]: I think we are not testing this.

**Review by dveeden — APPROVED** [review]
I think this is all good and I will merge this later today or tomorrow morning unless there are objections.

I had a look at the test coverage and there are a few very minor gaps that we could probably fix easily.

File: mysql/mariadb_gtid_test.go (hunk 3 above)
Comment (dveeden) [reply mysql/mariadb_gtid_test.go:111]: Maybe also something like `4294967295-1-2` and `4294967296-1-2`. I think this is `<uint32>-<uint32>-<uint64>`.

And I assume `1e5-1-2` doesn't work...

Comment (triplefighter7619) [reply mysql/mariadb_gtid_test.go:111]: Added test cases for `"x-1-1"` (non-numeric domain), `"0-x-1"` (non-numeric server), `"1e5-1-2"` (scientific notation), plus uint32/uint64 boundary values (`4294967295-1-2` valid, `4294967296-1-2` overflow error, `0-1-18446744073709551616` overflow error).

File: mysql/mariadb_gtid.go (hunk 4 above)
Comment (triplefighter7619) [reply mysql/mariadb_gtid.go:133]: Added `TestMariaDBGTIDSetForwardError` — the forward() error path is unreachable through normal `AddSet()`/`Update()` usage since the map key guarantees domain match, but tested via injected mismatch for completeness. Also added `Update()` with invalid strings (`"x-1-1"`, `"0-1--1"`).

File: mysql/mariadb_gtid.go
Comment (triplefighter7619) [reply mysql/mariadb_gtid.go:191]: Added: `mariaSet.Equal(mysqlSet)` → `false` (tests the type assertion `!ok` branch).

File: mysql/mariadb_gtid.go (hunk 5 above)
Comment (triplefighter7619) [reply mysql/mariadb_gtid.go:201]: Added: `{"1-1-1", "2-1-1", false}` — same length sets but different domain IDs, tests the `s.Sets[domainID]` not-found branch.

File: mysql/mariadb_gtid.go (hunk 6 above)
Comment (triplefighter7619) [reply mysql/mariadb_gtid.go:215]: Added: `mariaSet.Contain(mysqlSet)` → `false` (type assertion branch), and `{"1-1-1", "2-1-1", false}` (domain not found branch).
